src/App.tsx

```diff
@@ -1,29 +1,111 @@
 import React from 'react'
 import { Provider } from 'react-redux'
 import { ThemeContext, getTheme } from 'react-native-material-ui'
-import { Navigator } from 'src/NavigatorContainer'
 import { JolocomTheme } from 'src/styles/jolocom-theme'
 import { initStore } from './store'
+import { genericActions, accountActions, navigationActions } from 'src/actions'
+import { Linking, BackHandler, StatusBar } from 'react-native'
+import { NavigationActions } from 'react-navigation'
+import { withLoading, withErrorHandling } from './actions/modifiers'
+import { RoutesContainer } from './routes'
+
+import { AppLoading } from './ui/generic/appLoading'
+import { useScreens } from 'react-native-screens'
+import { routeList } from './routeList'
+useScreens()
 
 let store: ReturnType<typeof initStore>
 
-const App = () => {
-  // only init store once, or else Provider complains (especially on 'toggle
-  // inspector')
-  //
-  // but it needs to be done only when a new App is
-  // instantiated because otherwise the overrides at the top of index.ts will
-  // have not been excuted yet (while files are being imported) and initStore
-  // triggers creation of BackendMiddleware which needs those
-  if (!store) store = initStore()
-
-  return (
-    <ThemeContext.Provider value={getTheme(JolocomTheme)}>
-      <Provider store={store}>
-        <Navigator dispatch={store.dispatch} />
-      </Provider>
-    </ThemeContext.Provider>
-  )
-}
+export default class App extends React.PureComponent<{}> {
+  private navigator: any
+
+  constructor(props: {}) {
+    super(props)
+    // only init store once, or else Provider complains (especially on 'toggle
+    // inspector')
+    //
+    // but it needs to be done only when a new App is
+    // instantiated because otherwise the overrides at the top of index.ts will
+    // have not been excuted yet (while files are being imported) and initStore
+    // triggers creation of BackendMiddleware which needs those
+    if (!store) store = initStore()
+  }
+
+  /**
+   * should return false if app exit is desired
+   */
+  private navigateBack = () => {
+    if (!this.navigator) return
+
+    let curState = this.navigator.state.nav
+    while (true) {
```

Comment:
Ah yeah I will add a comment there. I actually commented about it in the commit message for 3e9e35accc08
```
    App.tsx: fix back button handler after routes change

    - exit when routeName === Claims, instead of specific index etc
    - use an arrow function for capturing this
```

Basically the check for whether we should exit the app or not was checking that the navigation state is a specific index at a specific depth, but it should actually check whether the current route is a specific route. The original check was invalid after I re-arranged the routes to use another layer (the `SwitchNavigator`), so I rewrote it to be route config agnostic, just cares that we are currently on the `Claims` route, regardless of where that is in the routes tree.

Reply:
As for the `while (true)`, it's because the actual check needs to grab stuff from `curState`, I tried to write it as a `do...while` and as a `for()` but they look/feel very clunky :D

Reply:
Ok rethinking it, I should be checking that we are on the first tab of Home, in case we decide to change what the first tab is. Of course we actually change 'Home` then this will need to be updated anyway

Reply:
Ah to address another concern, whether or not a loop is necessary at all: yes it is depending on how many levels deep we have to go to find the current route (again, depending on the routes configuration), so I'm using a loop just to not have to keep fixing this function every time there is a change to routes (already done so twice :D)

Reply:
Ok, another update on this: we don't even need to handle back button like that! the behavior being implemented is the default for react-navigation..... I mean if you have the routes set up right then if it ends up popping the last route from the stack it will close down the app :smile: I was just "fixing" up the code without even checking why we need it in the first place

https://reactnavigation.org/docs/en/custom-android-back-button-handling.html